### PR TITLE
Added Chainrule

### PR DESCRIPTION
**Pull request by MetalMiner**
See Issue #471

### CODE REVIEW

**Comment by milczarekIT** [comment]
Hi @MetalMiner,
You are using old version of code. Build fails due to compilation error.
Please merge current develop branch from this repo to your branch.

**Comment by MetalMiner** [comment]
Hi @milczarekIT ,

current development branch has now been merged and test has been updated.

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core.trading.rules.helper;
+
+import org.ta4j.core.Rule;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+public class ChainLink implements Serializable {
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java:35]: Please add javadoc. This class is not self-descriptive

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    private Rule rule;
+    private int threshold = 0;
+
+    /**
+     * Threshold is the number of bars the provided rule has to be satisfied after the preceding rule
+     *
+     * @param rule
+     * @param threshold
+     */
+    public ChainLink(Rule rule, int threshold) {
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java:45]: Please add description after @param rule/threshold

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Arrays;
+import java.util.LinkedList;
+
+public class ChainRule extends AbstractRule {
+    Rule initialRule = null;
+    LinkedList<ChainLink> rulesInChain = new LinkedList<>();
+
+    /**
+     * A chainrule has an initial rule that has to be satisfied before chain links are evaluated.
+     * If the initial rule is satisfied every rule of chain link has to be satisfied within a specified amount of bars (threshold).
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java]: Please move javadoc above the class and add descriptions for the @param initialRule/chainLinks

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import org.ta4j.core.Rule;
+import org.ta4j.core.TradingRecord;
+import org.ta4j.core.trading.rules.helper.ChainLink;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+
+public class ChainRule extends AbstractRule {
+    Rule initialRule = null;
+    LinkedList<ChainLink> rulesInChain = new LinkedList<>();
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java:40]: How about making this fileds `private final` ? Initial Rule should not be initialized with null but always via constructor. After that null check in line 54 would not be necessary

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        this.initialRule = initialRule;
+        rulesInChain.addAll(Arrays.asList(chainLinks));
+    }
+
+    @Override
+    public boolean isSatisfied(int index, TradingRecord tradingRecord) {
+        int lastRuleWasSatisfiedAfterBars = 0;
+        int startIndex = index;
+
+        if (initialRule == null || !initialRule.isSatisfied(index, tradingRecord)) {
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java]: avoid `initialRule` being `null` , please see my comment on top

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java (hunk 2 above)
Comment (team172011) [reply ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java:45]: also clarify if the current index is included or not e.g. if `threshold = 3` do you mean 

- this, last-1, last-2, last-3 **or** 
- this, last-1, last-2

File: ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        OverIndicatorRule overIndicatorRule = new OverIndicatorRule(indicator, 7);
+        IsEqualRule isEqualRule = new IsEqualRule(indicator, 5);
+        chainRule = new ChainRule(underIndicatorRule, new ChainLink(overIndicatorRule, 3), new ChainLink(isEqualRule, 2));
+    }
+
+    @Test
+    public void isSatisfied() {
+        assertFalse(chainRule.isSatisfied(0));
+        assertTrue(chainRule.isSatisfied(4));
+        assertTrue(chainRule.isSatisfied(6));
```
Comment (team172011) [inline ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java:56]: why should ChainRule return true for index `6` ? 
   
index 6 is 2

6, 5, 8, 5, 1, 10, **2**

- initial rule:         **2** < 5 true
- overIndicator (3):    **10**  > 7  || **1**  > 7 || **5** > 7 true 
- isEqualRule (2):       **2** == 5 || **10** == 5 **false**

File: ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+            for (int i = 0; i <= link.getThreshold(); i++) {
+                int resultingIndex = startIndex - i;
+                if (resultingIndex < 0) {
+                    break;
+                }
+
+                satisfiedWithinThreshold = link.getRule().isSatisfied(resultingIndex, tradingRecord);
+
+                if (satisfiedWithinThreshold == true) {
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java:75]: `== true` check is redundant

File: ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java (hunk 6 above)
Comment (MetalMiner) [reply ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java:56]: Hi @team172011 

there is a missunderstanding how this code works:

- initial rule: 2 < 5 true: This is an easy one, the rule is satisfied immediately and the next rule in the chain starts with the **same** index
- overIndicator (3): is therefore true at index 5, next rule starts on index 5
- isEqualRule (2): is therefore true on index 3 ;)

Comment (team172011) [reply ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java:56]: okay, this way this rule makes more sense :)

**Comment by milczarekIT** [comment]
This merge cause develop fails: 

```
[WARNING] Missing header in: /home/travis/build/ta4j/ta4j/ta4j-core/src/main/java/org/ta4j/core/trading/rules/ChainRule.java
[WARNING] Missing header in: /home/travis/build/ta4j/ta4j/ta4j-core/src/main/java/org/ta4j/core/trading/rules/helper/ChainLink.java
[WARNING] Missing header in: /home/travis/build/ta4j/ta4j/ta4j-core/src/test/java/org/ta4j/core/trading/rules/ChainRuleTest.java
```

@MetalMiner , @team172011  - could you look at this?
https://travis-ci.org/ta4j/ta4j/builds/656618964?utm_medium=notification&utm_source=github_status

**Pull request merged**